lib/src/model/challenge/challenge_repository.dart

```diff
@@ -82,3 +102,109 @@ class ChallengeRepository {
     }
   }
 }
+
+@Riverpod(keepAlive: true)
+class Challenges extends _$Challenges {
+  StreamSubscription<SocketEvent>? _subscription;
+
+  late SocketClient _socketClient;
+
+  @override
+  Future<ChallengesList> build() async {
+    _socketClient = ref.watch(socketPoolProvider).open(Uri(path: '/socket/v5'));
+
+    _subscription?.cancel();
+    _subscription = _socketClient.stream.listen(_handleSocketEvent);
+
+    ref.onDispose(() {
+      _subscription?.cancel();
+    });
+
+    return ref.read(challengeRepositoryProvider).list();
+  }
+
+  Future<GameFullId?> accept(ChallengeId id) async {
+    final repo = ref.read(challengeRepositoryProvider);
+    _stateRemove(id);
+
+    return repo
+        .accept(id)
+        .then((_) => repo.show(id).then((challenge) => challenge.gameFullId));
+  }
+
+  Future<void> decline(ChallengeId id) async {
+    final repo = ref.read(challengeRepositoryProvider);
+    _stateRemove(id);
+    return repo.decline(id);
+  }
+
+  void cancel(ChallengeId id) {
+    ref.read(challengeRepositoryProvider).cancel(id);
+  }
+
+  void _handleSocketEvent(SocketEvent event) {
+    if (event.topic != 'challenges') return;
+
+    final listPick = pick(event.data).required();
+    final inward = listPick('in').asListOrEmpty(Challenge.fromPick);
+    final outward = listPick('out').asListOrEmpty(Challenge.fromPick);
+
+    final prevIds = state.value?.inward.map((element) => element.id) ?? [];
+    // find any challenges that weren't in the inward list before
+    inward
+        .map((element) => element.id)
+        .where((id) => !prevIds.contains(id))
+        .map((id) => inward.firstWhere((element) => element.id == id))
+        .forEach(_notifyUser);
+
+    state = AsyncValue.data((inward: inward.lock, outward: outward.lock));
+  }
+
+  void _stateRemove(ChallengeId id) {
+    final inward = state.value!.inward;
+    final outward = state.value!.outward;
+    final challengeIn =
+        inward.firstWhereOrNull((challenge) => challenge.id == id);
+    final challengeOut =
+        outward.firstWhereOrNull((challenge) => challenge.id == id);
+
+    final newInward = challengeIn == null ? inward : inward.remove(challengeIn);
+    final newOutward =
+        challengeOut == null ? outward : outward.remove(challengeOut);
+    state = AsyncValue.data((inward: newInward, outward: newOutward));
+  }
+
+  void _notifyUser(Challenge challenge) {
+    final l10n = ref.read(l10nProvider).strings;
+
+    ref.read(localNotificationServiceProvider).show(
+          ChallengeNotification(
+            challenge,
+            l10n,
+            onPressed: (action, id) {
+              switch (action) {
+                case ChallengeNotificationAction
+                      .accept: // accept the game and open board
+                  final repo = ref.read(challengesProvider.notifier);
```

Comment:
As I explained in the other comment I don't think the challenges notifier should be used here.

It is better to use the challenge repository directly to make a server request, or use a dedicated service.

The provider (and associated notifier) is instantiated from the view (using ref.watch) and should only be used as a view controller (the term is perhaps not accurate but you get the idea).

I think this will simplify things a lot, since you don't need anymore to pass a message from isolates. The view needs only to be updated when the app is in foreground, and that should be done with the websocket only because the server is the source of truth. We don't want to modify the state locally from a user action, but just listen to server to update view state.